netket/models/autoreg.py

```diff
@@ -158,8 +186,73 @@ def setup(self):
             for i in range(self.layers)
         ]
 
-    def conditionals(self, inputs: Array, cache: PyTree) -> Tuple[Array, PyTree]:
-        return _conditionals(self, inputs, cache)
+    def _conditional(self, inputs: Array, index: int) -> Array:
+        return _conditionals(self, inputs)[:, index, :]
+
+    def conditionals(self, inputs: Array) -> Array:
+        return _conditionals(self, inputs)
+
+    def __call__(self, inputs: Array) -> Array:
+        return _call(self, inputs)
+
+
+class ARNNConv2D(ARNN):
+    """Autoregressive neural network with 2D convolution layers."""
+
+    layers: int
+    """number of layers."""
+    features: Union[Iterable[int], int]
+    """number of features in each layer. If a single number is given,
+    all layers except the last one will have the same number of features."""
+    kernel_size: Tuple[int, int]
+    """shape of the convolutional kernel `(h, w)`. Typically, `h = w // 2 + 1`."""
+    kernel_dilation: Tuple[int, int] = (1, 1)
+    """a sequence of 2 integers, giving the dilation factor to
+    apply in each spatial dimension of the convolution kernel (default: 1)."""
+    activation: Callable[[Array], Array] = jax.nn.selu
+    """the nonlinear activation function between hidden layers (default: selu)."""
+    use_bias: bool = True
+    """whether to add a bias to the output (default: True)."""
+    dtype: DType = jnp.float64
+    """the dtype of the weights (default: float64)."""
+    precision: Any = None
+    """numerical precision of the computation, see `jax.lax.Precision` for details."""
+    kernel_init: NNInitFunc = default_kernel_init
+    """initializer for the weights."""
+    bias_init: NNInitFunc = zeros
+    """initializer for the biases."""
+
+    def setup(self):
```

Comment:
is this supposed to be a public function that can be called freely by the user?

Reply:
No, `setup` is called by Flax to do some initialization. And we cannot add an underscore before the name because it's named by Flax.